File: src-tauri/src/route/admin/rag_repositories.rs
use aide::{
    axum::{ApiRouter, routing::{delete_with, get_with, post_with, put_with}},
};

use crate::api::rag_repositories::{
    create_rag_repository, delete_rag_repository, download_rag_database_from_repository,
    get_rag_repository, list_rag_repositories, list_rag_repository_databases,
    test_rag_repository_connection, update_rag_repository,
};

pub fn admin_rag_repository_routes() -> ApiRouter {
    ApiRouter::new()
        // RAG Repository routes
        .api_route(
            "/rag-repositories",
            get_with(list_rag_repositories, |op| {
                op.description("List all RAG repositories")
                    .id("Admin.listRAGRepositories")
                    .tag("admin")
            })
        )
        .api_route(
            "/rag-repositories",
            post_with(create_rag_repository, |op| {
                op.description("Create a new RAG repository")
                    .id("Admin.createRAGRepository")
                    .tag("admin")
            })
        )
        .api_route(
            "/rag-repositories/{repository_id}",
            get_with(get_rag_repository, |op| {
                op.description("Get a specific RAG repository")
                    .id("Admin.getRAGRepository")
                    .tag("admin")
            })
        )
        .api_route(
            "/rag-repositories/{repository_id}",
            put_with(update_rag_repository, |op| {
                op.description("Update a RAG repository")
                    .id("Admin.updateRAGRepository")
                    .tag("admin")
            })
        )
        .api_route(
            "/rag-repositories/{repository_id}",
            delete_with(delete_rag_repository, |op| {
                op.description("Delete a RAG repository")
                    .id("Admin.deleteRAGRepository")
                    .tag("admin")
            })
        )
        .api_route(
            "/rag-repositories/{repository_id}/test-connection",
            post_with(test_rag_repository_connection, |op| {
                op.description("Test RAG repository connection")
                    .id("Admin.testRAGRepositoryConnection")
                    .tag("admin")
            })
        )
        .api_route(
            "/rag-repositories/{repository_id}/databases",
            get_with(list_rag_repository_databases, |op| {
                op.description("List databases for a RAG repository")
                    .id("Admin.listRAGRepositoryDatabases")
                    .tag("admin")
            })
        )
        .api_route(
            "/rag-repositories/download-database",
            post_with(download_rag_database_from_repository, |op| {
                op.description("Download RAG database from repository")
                    .id("Admin.downloadRAGDatabaseFromRepository")
                    .tag("admin")
            })
        )
}

File: src/store/admin/ragRepositories.ts
import { create } from 'zustand'
import { subscribeWithSelector } from 'zustand/middleware'
import { ApiClient } from '../../api/client.ts'
import {
  CreateRAGRepositoryRequest,
  DownloadRAGDatabaseFromRepositoryRequest,
  RAGDatabase,
  RAGRepository,
  UpdateRAGRepositoryRequest,
} from '../../types'

interface AdminRAGRepositoriesState {
  // Data
  repositories: RAGRepository[]
  availableDatabasesByRepository: Record<string, RAGDatabase[]>

  // Loading states
  loading: boolean
  creating: boolean
  updating: boolean
  deleting: boolean
  loadingDatabases: Record<string, boolean>
  testingConnection: Record<string, boolean>
  downloading: Record<string, boolean>

  // Error state
  error: string | null
}

export const useAdminRAGRepositoriesStore = create<AdminRAGRepositoriesState>()(
  subscribeWithSelector(
    (): AdminRAGRepositoriesState => ({
      // Initial state
      repositories: [],
      availableDatabasesByRepository: {},
      loading: false,
      creating: false,
      updating: false,
      deleting: false,
      loadingDatabases: {},
      testingConnection: {},
      downloading: {},
      error: null,
    }),
  ),
)

// Repository actions
export const loadAllRAGRepositories = async (): Promise<void> => {
  try {
    useAdminRAGRepositoriesStore.setState({ loading: true, error: null })
    const response = await ApiClient.Admin.listRAGRepositories({})
    useAdminRAGRepositoriesStore.setState({
      repositories: response.repositories,
      loading: false,
    })
  } catch (error) {
    useAdminRAGRepositoriesStore.setState({
      error:
        error instanceof Error
          ? error.message
          : 'Failed to load RAG repositories',
      loading: false,
    })
    throw error
  }
}

export const createNewRAGRepository = async (
  repository: CreateRAGRepositoryRequest,
): Promise<RAGRepository> => {
  try {
    useAdminRAGRepositoriesStore.setState({ creating: true, error: null })
    const newRepository = await ApiClient.Admin.createRAGRepository(repository)
    useAdminRAGRepositoriesStore.setState(state => ({
      repositories: [...state.repositories, newRepository],
      creating: false,
    }))
    return newRepository
  } catch (error) {
    useAdminRAGRepositoriesStore.setState({
      error:
        error instanceof Error
          ? error.message
          : 'Failed to create RAG repository',
      creating: false,
    })
    throw error
  }
}

export const updateRAGRepository = async (
  id: string,
  repository: UpdateRAGRepositoryRequest,
): Promise<void> => {
  try {
    useAdminRAGRepositoriesStore.setState({ updating: true, error: null })
    const updatedRepository = await ApiClient.Admin.updateRAGRepository({
      repository_id: id,
      ...repository,
    })
    useAdminRAGRepositoriesStore.setState(state => ({
      repositories: state.repositories.map(r =>
        r.id === id ? updatedRepository : r,
      ),
      updating: false,
    }))
  } catch (error) {
    useAdminRAGRepositoriesStore.setState({
      error:
        error instanceof Error
          ? error.message
          : 'Failed to update RAG repository',
      updating: false,
    })
    throw error
  }
}

export const deleteRAGRepository = async (id: string): Promise<void> => {
  try {
    useAdminRAGRepositoriesStore.setState({ deleting: true, error: null })
    await ApiClient.Admin.deleteRAGRepository({ repository_id: id })
    useAdminRAGRepositoriesStore.setState(state => ({
      repositories: state.repositories.filter(r => r.id !== id),
      availableDatabasesByRepository: Object.fromEntries(
        Object.entries(state.availableDatabasesByRepository).filter(
          ([repositoryId]) => repositoryId !== id,
        ),
      ),
      deleting: false,
    }))
  } catch (error) {
    useAdminRAGRepositoriesStore.setState({
      error:
        error instanceof Error
          ? error.message
          : 'Failed to delete RAG repository',
      deleting: false,
    })
    throw error
  }
}

export const testRAGRepositoryConnection = async (
  id: string,
): Promise<void> => {
  try {
    useAdminRAGRepositoriesStore.setState(state => ({
      testingConnection: { ...state.testingConnection, [id]: true },
      error: null,
    }))

    await ApiClient.Admin.testRAGRepositoryConnection({
      repository_id: id,
    })

    useAdminRAGRepositoriesStore.setState(state => ({
      testingConnection: { ...state.testingConnection, [id]: false },
    }))
  } catch (error) {
    useAdminRAGRepositoriesStore.setState(state => ({
      error:
        error instanceof Error
          ? error.message
          : 'Repository connection test failed',
      testingConnection: { ...state.testingConnection, [id]: false },
    }))
    throw error
  }
}

// Database actions
export const loadAvailableDatabasesFromRepository = async (
  repositoryId: string,
): Promise<void> => {
  try {
    useAdminRAGRepositoriesStore.setState(state => ({
      loadingDatabases: { ...state.loadingDatabases, [repositoryId]: true },
      error: null,
    }))

    const databases = await ApiClient.Admin.listRAGRepositoryDatabases({
      repository_id: repositoryId,
    })

    useAdminRAGRepositoriesStore.setState(state => ({
      availableDatabasesByRepository: {
        ...state.availableDatabasesByRepository,
        [repositoryId]: databases,
      },
      loadingDatabases: { ...state.loadingDatabases, [repositoryId]: false },
    }))
  } catch (error) {
    useAdminRAGRepositoriesStore.setState(state => ({
      error:
        error instanceof Error
          ? error.message
          : 'Failed to load available RAG databases',
      loadingDatabases: { ...state.loadingDatabases, [repositoryId]: false },
    }))
    throw error
  }
}

export const downloadRAGDatabaseFromRepository = async (
  request: DownloadRAGDatabaseFromRepositoryRequest,
): Promise<void> => {
  try {
    const downloadKey = `${request.repository_id}-${request.database_id}`
    useAdminRAGRepositoriesStore.setState(state => ({
      downloading: { ...state.downloading, [downloadKey]: true },
      error: null,
    }))

    await ApiClient.Admin.downloadRAGDatabaseFromRepository(request)

    useAdminRAGRepositoriesStore.setState(state => ({
      downloading: { ...state.downloading, [downloadKey]: false },
    }))
  } catch (error) {
    const downloadKey = `${request.repository_id}-${request.database_id}`
    useAdminRAGRepositoriesStore.setState(state => ({
      error:
        error instanceof Error
          ? error.message
          : 'Failed to download RAG database',
      downloading: { ...state.downloading, [downloadKey]: false },
    }))
    throw error
  }
}

// Utility actions
export const clearRAGRepositoriesError = (): void => {
  useAdminRAGRepositoriesStore.setState({ error: null })
}

export const findRAGRepositoryById = (
  id: string,
): RAGRepository | undefined => {
  return useAdminRAGRepositoriesStore
    .getState()
    .repositories.find(r => r.id === id)
}

export const searchRAGRepositories = (
  repositories: RAGRepository[],
  query: string,
): RAGRepository[] => {
  if (!query.trim()) return repositories

  const searchTerm = query.toLowerCase()
  return repositories.filter(
    repository =>
      repository.name.toLowerCase().includes(searchTerm) ||
      repository.description?.toLowerCase().includes(searchTerm) ||
      repository.url.toLowerCase().includes(searchTerm),
  )
}

export const searchAvailableRAGDatabases = (
  databases: RAGDatabase[],
  query: string,
): RAGDatabase[] => {
  if (!query.trim()) return databases

  const searchTerm = query.toLowerCase()
  return databases.filter(
    database =>
      database.name.toLowerCase().includes(searchTerm) ||
      database.alias.toLowerCase().includes(searchTerm) ||
      database.description?.toLowerCase().includes(searchTerm),
  )
}
